### PR TITLE
Files duplicate basenames

### PR DESCRIPTION
**Pull request by rienafairefr**
As discussed this morning.

The heart of the matter is `iotlabcli.helpers.FilesDict` which is the class responsible for storing experiment files before they are shipped off in multipart. It couldn't handle two different files with the same basename because the indexing key is the basename and it errors when modifying an element to a different value.

I've modified the add_file method to be able to handle duplicated basenames with different content, modifying the file name to prepend the MD5 hash. 
I've modified add_files_from_dict so that this method modifies the input files_dict (containing the association between nodes and firmwares, etc), that might be considered a feature (we can thus e.g. print the submission with the modified firmware names in it before being sent).

In the process, I've migrated that part to pytest (doctests included), I should probably make that (
47551ee) a fully separate commit

### CODE REVIEW

**Comment by fsaintma** [comment]
@cladmi should you review this PR and say us if is it ok for you ? After we will merge and publish a new cli-tools version and deploy it on testbed.
Thanks

File: iotlabcli/helpers.py [hunk 1]
```diff
@@ -188,15 +178,30 @@ def __setitem__(self, key, val):
...
+        self[key] = read_file(file_path, 'b')
+        return key
 
     def add_files_from_dict(self, keys, files_dict):
         """Add 'keys' files from 'files_dict' if present."""
         for key in keys:
-            self.add_file(files_dict.get(key, None))
+            file_element = files_dict.get(key, None)
+            if file_element is not None:
+                files_dict[key] = self.add_file(file_element)
```
Comment (cladmi) [inline iotlabcli/helpers.py]: Why do you need to update `files_dict` there ?

From what I remember it is supposed to be only used as an input.
This would remove the special handling done here about checking if the file is empty, and remove the need for a special test to cover if `file_path` is None in `add_file`.

Comment (cladmi) [reply iotlabcli/helpers.py]: Also, in there, you are replacing the content of `files_dict[key] == absolute_file_path` by `files_dict[key] == actual_key == (basename or md5_basename)` so not sure it is a wanted side effect.

File: iotlabcli/helpers.py [hunk 2]
```diff
@@ -188,15 +178,30 @@ def __setitem__(self, key, val):
...
+        then prefix with short hash is used
+        if None do nothing
+        """
         if file_path is None:
-            return
-        self[os.path.basename(file_path)] = read_file(file_path, 'b')
+            return None
+        key = os.path.basename(file_path)
+        value = read_file(file_path, 'b')
+        if key in self and self[key] != value:
```
Comment (cladmi) [inline iotlabcli/helpers.py]: This is duplicating the check done in `__setitem__` in a pythonic way, would be better to do something like:

```
try:
    self[key] = value
except ValueError:
    key = '{hash}_{path}`.format(hash=md5(value), path=key)
    self[key] = value
```

    self['{hash}_{path}`.format(hash=md5

File: iotlabcli/helpers.py [hunk 3]
```diff
@@ -188,15 +178,30 @@ def __setitem__(self, key, val):
...
+            # use md5 hash as prefix to handle duplicated basenames
+            # with different contents
+            key = md5(value) + '_' + key
```
Comment (cladmi) [inline iotlabcli/helpers.py]: Please no Java string handling :p

File: iotlabcli/helpers.py [hunk 4]
```diff
@@ -188,15 +178,30 @@ def __setitem__(self, key, val):
...
+        self[key] = read_file(file_path, 'b')
```
Comment (cladmi) [inline iotlabcli/helpers.py]: You are re-reading the file here a second time when you already have `value`.

File: iotlabcli/experiment.py [hunk 5]
```diff
@@ -64,14 +64,14 @@ def submit_experiment(api, name, duration,  # pylint:disable=too-many-arguments
 
     exp_files = helpers.FilesDict()
     for res_dict in resources:
-        experiment.add_exp_resources(res_dict)
         exp_files.add_files_from_dict(NODES_ASSOCIATIONS_FILE_ASSOCS, res_dict)
+        experiment.add_exp_resources(res_dict)
```
Comment (cladmi) [inline iotlabcli/experiment.py]: Why are these out of the previous commit ? The commit message does not explain what you are doing here.

File: iotlabcli/tests/helpers_test.py [hunk 6]
```diff
@@ -102,53 +104,71 @@ def fake_cmd():
...
+# Test FilesDict class
 
-    def test_no_overwrite(self):
-        """Test FilesDict values overriding."""
-        file_dict = helpers.FilesDict()
-        file_dict['a'] = 1
-        # Can re-add same value
-        file_dict['a'] = 1
+@pytest.fixture(scope='function')
+def file_dict():
```
Comment (cladmi) [inline iotlabcli/tests/helpers_test.py]: Why using a fixture here, it is not as if it was as a prepared mocked environment, the tests are actually testing the class so make sense to initialize it in the test.

File: iotlabcli/experiment.py (hunk 5 above)
Comment (rienafairefr) [reply iotlabcli/experiment.py]: add_files_from_dict modifies res_dict (because we are modifying the file name), that's why. Will update the commit message or squash

File: iotlabcli/helpers.py (hunk 2 above)
Comment (rienafairefr) [reply iotlabcli/helpers.py]: I agree

File: iotlabcli/helpers.py (hunk 3 above)
Comment (rienafairefr) [reply iotlabcli/helpers.py]: I've been polluted by working on testbed-manager, argh ^^

File: iotlabcli/helpers.py (hunk 1 above)
Comment (rienafairefr) [reply iotlabcli/helpers.py]: the file_path None as marker for "do nothing" was already there before I modified, I kept it out of caution, would have done differently

Comment (rienafairefr) [reply iotlabcli/helpers.py]: Yes, I'm modifying files_dict. Which smelled bad to me, and I would love to do it differently. Might completely scrap the add_files_from_dict method actually

File: iotlabcli/tests/helpers_test.py (hunk 6 above)
Comment (rienafairefr) [reply iotlabcli/tests/helpers_test.py]: it's either a fixture and file_dict everywhere or repeating helpers.FilesDict() everywhere, either way there's repetition. Just going pytest all the way I guess.

File: iotlabcli/helpers.py (hunk 4 above)
Comment (rienafairefr) [reply iotlabcli/helpers.py]: You're right, will be removed

File: iotlabcli/helpers.py (hunk 1 above)
Comment (rienafairefr) [reply iotlabcli/helpers.py]: It's been fixed, the modified dict is returned to the caller.

File: iotlabcli/helpers.py (hunk 2 above)
Comment (rienafairefr) [reply iotlabcli/helpers.py]: Fixed

**Comment by cladmi** [comment]
I lost on this one, I put it on my tomorrow plans.
